[#190] 댓글: 댓글 화면 및 입력 구현

jwonyLee opened the pull request:
## 개요

<img src="https://user-images.githubusercontent.com/15073405/142733531-af03bc63-2132-464f-861c-095e2cfd3375.gif" width="50%" />

댓글 화면 구현

## 작업 사항

- TextViewWithSideButtonsView 구현
  - `글쓰기 > 댓글` 화면 하단에 들어가는 댓글 입력 뷰
- 댓글 화면 구현
  - 테이블 뷰와 TextviewWithSideButtonsView를 이용해 구현했습니다.

## 예외 사항

- 실제 데이터 바인딩
  - 댓글 수정, 삭제, 저장 기능 X
- iOS 13, 14에서 댓글 수정 모드로 진입할 때 [수정/편집 완료] 텍스트가 변경되면서 Label이 늘어났다가 줄어드는 형태로 표시됩니다. 이 문제는 텍스트뷰에 키보드 포커싱이 있을 때 생기는 거 같은데, 해결 방안을 찾지 못했습니다. ㅠㅠ

## 테스트

설정 화면에서 확인할 수 있습니다.

### Linked Issue

close #190

jwonyLee committed:
[#190] view: TextViewWithSideButtonsView 구현
글쓰기>댓글 화면 하단에 들어가는 댓글 입력 뷰

jwonyLee committed:
[#190] view: 게시물>댓글 화면 구현
- 댓글 화면 구성을 위한 뷰 배치
- 키보드가 숨김/표시될 때 댓글 입력 뷰 오토레이아웃 변경
- TableView DataSource 구현

jwonyLee committed:
[#190] fix: TextViewWithSideButtonsView leftButton 호출 함수 추가
기존에는 leftButton을 탭하면 textView.text = nil로 설정했다. 이렇게 하면 이미 플레이스 홀더가 표시되고 있을 때 leftButton을 누르면 플레이스 홀더가 지워지는 현상이 있어서 placeholder를 설정하는 함수를 별도로 만들고 해당 함수를 호출하는 방식으로 변경했다.

jwonyLee committed:
[#190] test: 설정 화면에서 댓글 화면 표시

jwonyLee committed:
[#190] feat: TextView 최대 3줄까지 노출되게 변경

jwonyLee committed:
[#190] feat: 댓글 편집모드 구현

jwonyLee committed:
[#190] feat: 텍스트뷰의 포커싱이 바뀌었을 때 편집 모드 false로 변경

gustn3965 commented on ThingLog/View/TableVeiwCell/CommentTableCell.swift:
```diff
@@ -116,5 +138,55 @@ final class CommentTableCell: UITableViewCell {
+        isEditable ? modifyButton.setTitle("편집 완료", for: .normal) : modifyButton.setTitle("수정", for: .normal)
+        isEditable ? modifyButton.setTitleColor(SwiftGenColors.systemGreen.color, for: .normal) : modifyButton.setTitleColor(SwiftGenColors.primaryBlack.color, for: .normal)
+        isEditable ? deleteButton.setTitle("취소", for: .normal) : deleteButton.setTitle("삭제", for: .normal)
+    }
```
안에서 하는 방법도 더 간결하게 할 수 있는 것 같습니당 
```swift
    private func setEditMode() {
        textView.isEditable = isEditable
        textView.isUserInteractionEnabled = isEditable
        modifyButton.setTitle(isEditable ? "편집 완료" : "수정", for: .normal)
        modifyButton.setTitleColor(isEditable ? SwiftGenColors.systemGreen.color : SwiftGenColors.primaryBlack.color, for: .normal)
        deleteButton.setTitle(isEditable ? "취소" : "삭제", for: .normal)
    }
```

gustn3965 commented on ThingLog/View/TableVeiwCell/CommentTableCell.swift:
```diff
@@ -116,5 +138,55 @@ final class CommentTableCell: UITableViewCell {
+    private func dismissKeyboard() {
+        textView.resignFirstResponder()
+    }
+
```
지금, 수정 버튼을 누른다음, 툴바의 취소버튼을 누르면 `CommentViewController` 의 `commentInputView`가 계속 사라져있는 상태입니다. 

제가 생각나는 방법은, `dismissKeyBoard()`가 안에서 호출하는 코드를 클로저로 프로퍼티로 만들어서, 외부에서 주입하는 방법이 있는 것 같습니다.

gustn3965 commented on ThingLog/View/TextViewWithSideButtonsView.swift:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    /// leftButton을 탭하면 텍스트 뷰의 모든 텍스트를 지우고, 키보드를 내린다.
+    private func bindLeftButton() {
+        leftButton.rx.tap
+            .bind { [weak self] in
+                self?.textViewHeightConstraint?.constant = UIFont.Pretendard.body1.lineHeight
+                self?.setupPlaceholder()
+                self?.textView.resignFirstResponder()
+            }.disposed(by: disposeBag)
+    }
```
3줄 이상 입력하고 나서, 취소를 누르게 되면,  아래와같이 텍스트뷰가 삐져나오게 됩니다. 

![Pasted Graphic 8](https://user-images.githubusercontent.com/48749182/142753188-a51c9558-61e2-4964-aad0-3660bdd21d7f.jpg)

그래서 입력하고 있는 상태를 디버깅해보니,  여기서도 스택뷰 전체가 늘어나기 보다는 텍스트뷰만 삐져나온걸 볼 수 있습니다. 

![Pasted Graphic 7](https://user-images.githubusercontent.com/48749182/142753196-7021dfc3-8eec-45e3-a1aa-b9171e8f5aee.jpg)


그래서 저도 정확한 해결책은 모르겠지만, `stackView`의 `alignment`를 `center`로 하니 해결됐습니다.

gustn3965 commented on ThingLog/ViewController/Comment/CommentViewController+DataSource.swift:
```diff
@@ -0,0 +1,78 @@
+
+        cell.delegate = self
+        cell.textView.text = "테스트"
+        cell.modifyButton.rx.tap
+            .bind { [weak self] in
+                cell.isEditable.toggle()
+                cell.textView.isEditable ? cell.textView.becomeFirstResponder() : cell.textView.resignFirstResponder()
+                self?.hideCommentInputView(cell.isEditable)
+            }.disposed(by: cell.disposeBag)
+
+        cell.deleteButton.rx.tap
+            .bind { [weak self] in
+                if cell.isEditable {
+                    cell.isEditable.toggle()
+                    cell.textView.resignFirstResponder()
+                } else {
+                    // TODO: 삭제 기능
+                }
+                self?.hideCommentInputView(cell.isEditable)
+            }.disposed(by: cell.disposeBag)
```
* iOS 13, 14에서 댓글 수정 모드로 진입할 때 [수정/편집 완료] 텍스트가 변경되면서 Label이 늘어났다가 줄어드는 형태로 표시됩니다. 이 문제는 텍스트뷰에 키보드 포커싱이 있을 때 생기는 거 같은데, 해결 방안을 찾지 못했습니다. ㅠㅠ


 이 문제는 늘어났다가 줄어드는 형태보다는 애니메이션이 되고 있는 현상입니다. 그러면, 왜 애니메이션이 되고 있는지를 생각해볼 필요가 있어요.  
코드로직을 살펴보면, 분명 애니메이션적용한 코드는 키보드가 나타나거나 사라지는 경우에만 적용한 것 같지만요. 

좀 더 자세하게 살펴보면, 수정 버튼을 클릭할 경우 아래 로직이 이러한데, 
`isEditable`이 값이 변할 때 마다 `setEditMode` 메소드가 호출되어, 버튼뷰를 변경하고 있습니다. 
그리고 그런다음` textView`를 활성화하여, 키보드가 나타나도록 호출하는데, 

```swift
        cell.modifyButton.rx.tap
            .bind { [weak self] in
                cell.isEditable.toggle()
                cell.textView.isEditable ? cell.textView.becomeFirstResponder() : cell.textView.resignFirstResponder()
                self?.hideCommentInputView(cell.isEditable)
            }.disposed(by: cell.disposeBag)
```

키보드가 나타날 때  `setupCommentInputViewBottomConstraint()` 메소드를 호출하는데, 이 때 마지막 줄에 애니메이션이 되도록 하는 코드가 되어있죠. 뷰가 변형될 때는 시스템에서 특정 사이클을 돌면서 변형이 되는데, 애니메이션안에 `layoutIfneeded()`를 호출함으로써, 키보드만 애니메이션이 적용되는 것이 아닌, 그전 코드 ( 버튼 변경하는 코드 )들도 영향을 받을 수 있습니다.

그래서 해결방법은, 다양할텐데, `layoutIfneeded()`를 적절히 배치하는 것이겠죠. 원하는 부분만 애니메이션이 되도록 그전에는 미리 뷰가 다 업데이트되도록 한다던가, 등이 있겠죠. 

한번 어느정도 이해하고 나면 크게 어려운 부분은 아닙니다. 


추가적으로, 현재 `CommentViewController`가 `Notification`으로 옵저빙하고 있기 때문에,  `commentInputView`에 대한 키보드 뿐만 아니라, `commtentCell`에서의 키보드도 나타나거나 사라질때도 반응하게 됩니다.  문제가 있다는 코드가 아니라, 수정 삭제의 애니메이션을 해결하는 과정에서 놓칠 수도 있는 부분이라 말씀드려봅니다.

gustn3965 commented on ThingLog/ViewController/Comment/CommentViewController+setup.swift:
```diff
@@ -0,0 +1,104 @@
+            }.disposed(by: disposeBag)
+    }
+
+    /// CommentInputView의 하단 constant를 설정한다.
+    func setupCommentInputViewBottomConstraint(height: CGFloat, duration: TimeInterval) {
+        let bottomInset: CGFloat = view.safeAreaInsets.bottom
+        if height == 0 && bottomInset != 0 {
+            commentInputViewBottomConstraint?.constant = -height
+        } else {
```
비슷한 메소드들은 같은 곳에 위치시켜주면 좋을 것 같습니다. `setupTableViewBottomInset`은 `bind**` 메소드보다 위에 위치해있는 반면, 이 코드는 `bind**`메소드보다 아래에 위치해있습니다.  개인적으로 저는 `setup`과 `bind**` 메소드 사이에 `MARK`주석이 있으면 좀 더 보기 편할 것 같습니다.  🙏🏻

그리고 `ComemntViewController`와, `TextViewWithSideButtonsView`에서 `Properties` 와 `init`메소드 사이에 `// MARK` 주석을 넣어주시면 좀 더 보기 편할 것 같습니다. 🙏🏻 ( 저는 주로 `control + 6 `으로 이름을 봐서 대략적인 구조부터 보는 편이라서요 ㅎㅎ )

gustn3965 reviewed (commented):
고생하셨습니다. 아래 코멘트들 확인해주세요 ~ !

jwonyLee committed:
[#190] ref: 피드백 반영
- setEditMode() 수정
- TextViewWithSideButtonsView 정렬 변경
- iOS 13, 14에서 [수정/편집 완료] 텍스트가 늘어났다가 줄어드는 문제 해결
- 주석 추가
- 키보드 툴바 취소 시 댓글 편집 종료

jwonyLee commented:
피드백 반영했습니다!
- setEditMode() 수정
- TextViewWithSideButtonsView 정렬 변경
- iOS 13, 14에서 [수정/편집 완료] 텍스트가 늘어났다가 줄어드는 문제 해결
- 주석 추가
- 키보드 툴바 취소 시 댓글 편집 종료

gustn3965 reviewed (approved):
LGTM!

jwonyLee committed:
Merge branch 'develop' into view/190-comment
